haystack/nodes/prompt/prompt_node.py

```diff
@@ -0,0 +1,822 @@
+    For example, in the above example, the prompt_params are ["documents"] and the prompt_text is
+    "Please give a sentiment..."
+
+    The prompt_text contains a placeholder $documents. This variable will be filled in runtime with the non-keyword
+    or keyword argument `documents` passed to this PromptTemplate's fill() method.
+    """
+
+    def __init__(self, name: str, prompt_text: str, prompt_params: Optional[List[str]] = None):
+        super().__init__()
+        if not prompt_params:
+            # Define the regex pattern to match the strings after the $ character
+            pattern = r"\$([a-zA-Z0-9_]+)"
+            prompt_params = re.findall(pattern, prompt_text)
+
+        if prompt_text.count("$") != len(prompt_params):
+            raise ValueError(
+                f"Number of parameters in prompt text {prompt_text} for prompt template {name} "
+                f"does not match number of specified parameters {prompt_params}"
+            )
+
+        # use case when PromptTemplate is loaded from a YAML file, we need to start and end the prompt text with quotes
+        prompt_text = prompt_text.strip("'").strip('"')
+
+        t = Template(prompt_text)
+        try:
+            t.substitute(**{param: "" for param in prompt_params})
+        except KeyError as e:
+            raise ValueError(
+                f"Invalid parameter {e} in prompt text "
+                f"{prompt_text} for prompt template {name}, specified parameters are {prompt_params}"
+            )
+
+        self.name = name
+        self.prompt_text = prompt_text
+        self.prompt_params = prompt_params
+
+    def fill(self, *args, **kwargs) -> Dict[str, Any]:
+        """
+        Prepares the prompt text parameters from non-keyword and keyword arguments.
+
+        In the case of non-keyword arguments, the order of the arguments should match the left-to-right
+        order of appearance of the parameters in the prompt text. For example, if the prompt text is:
+        `Please come up with a question for the given context and the answer. Context: $documents;
+        Answer: $answers; Question:` then the first non-keyword argument will fill the $documents placeholder
+        and the second non-keyword argument will fill the $answers placeholder.
+
+        In the case of keyword arguments, the order of the arguments does not matter. Placeholders in the
+        prompt text are filled with the corresponding keyword argument.
+
+        :param args: non-keyword arguments to use for filling the prompt text
+        :param kwargs: keyword arguments to use for filling the prompt text
+        :return: a dictionary with the prompt text and the prompt parameters
+        """
+        template_dict = {}
+        # attempt to resolve args first
+        if args:
+            if len(args) != len(self.prompt_params):
+                logger.warning(
```

Comment:
I think we should continue to make sure not to use f-strings in our logging statements as we found that it can unintentionally affect performance as found in this PR https://github.com/deepset-ai/haystack/pull/3212. So instead something like 

```python
logger.warning("For %s, expected %s arguments, ...", self.name, self.prompt_params)
```
I think would be preferred.